File: packages/go/internal/seguranca/ataques/ataque_bytes_fixos.go
package ataques

import (
	"fmt"
	"strings"

	"criptografia/internal/seguranca"
)

// AtaqueBytesFixos gera vários tokens de textos diferentes e procura qualquer
// posição de byte que NUNCA muda.
type AtaqueBytesFixos struct {
	amostras int
}

// NovoAtaqueBytesFixos cria o ataque. Sem argumento, usa amostras = 30.
func NovoAtaqueBytesFixos(amostras ...int) *AtaqueBytesFixos {
	n := 30
	if len(amostras) > 0 {
		n = amostras[0]
	}
	return &AtaqueBytesFixos{amostras: n}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueBytesFixos) Nome() string {
	return "Bytes fixos entre tokens"
}

// Executar roda o ataque.
func (a *AtaqueBytesFixos) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	decodificados := [][]byte{}

	for i := 0; i < a.amostras; i++ {
		texto := fmt.Sprintf("TEXTO_VARIADO_%d", i) +
			strings.Repeat(string(rune('A'+(i%26))), i%10)
		token, err := alvo.Encrypt([]byte(texto))
		if err != nil {
			continue
		}
		decodificado, err := alvo.Base64urlDecode(token[len(alvo.Prefixo()):])
		if err != nil {
			continue
		}
		decodificados = append(decodificados, decodificado)
	}

	if len(decodificados) == 0 {
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: false,
			Severidade: seguranca.SevInfo,
			Detalhes:   "Nenhum token pôde ser analisado",
		}, nil
	}

	tamanhoMinimo := len(decodificados[0])
	for _, d := range decodificados {
		if len(d) < tamanhoMinimo {
			tamanhoMinimo = len(d)
		}
	}

	posicoesFixas := []int{}
	for i := 0; i < tamanhoMinimo; i++ {
		valores := map[byte]bool{}
		for _, d := range decodificados {
			valores[d[i]] = true
		}
		if len(valores) == 1 {
			posicoesFixas = append(posicoesFixas, i)
		}
	}

	if len(posicoesFixas) > 0 {
		limite := min(10, len(posicoesFixas))
		partes := make([]string, limite)
		for i := 0; i < limite; i++ {
			partes[i] = fmt.Sprint(posicoesFixas[i])
		}
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevAlta,
			Detalhes: fmt.Sprintf("%d posição(ões) de byte fixas em %d tokens: %s",
				len(posicoesFixas), a.amostras, strings.Join(partes, ", ")),
			Dados: map[string]any{"posicoes": posicoesFixas},
		}, nil
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: false,
		Severidade: seguranca.SevInfo,
		Detalhes:   fmt.Sprintf("0 de %d posições fixas em %d tokens", tamanhoMinimo, a.amostras),
	}, nil
}

File: packages/rust/src/seguranca/ataques/ataque_bytes_fixos.rs
//! Procura posições de byte que NUNCA mudam entre tokens — âncoras que um
//! atacante poderia usar pra recuperar a chave.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};

pub struct AtaqueBytesFixos {
    pub amostras: usize,
}

impl AtaqueBytesFixos {
    pub fn new() -> Self {
        AtaqueBytesFixos { amostras: 30 }
    }
}

impl Default for AtaqueBytesFixos {
    fn default() -> Self {
        Self::new()
    }
}

impl Ataque for AtaqueBytesFixos {
    fn nome(&self) -> String {
        "Bytes fixos entre tokens".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let mut decodificados: Vec<Vec<u8>> = Vec::new();

        for i in 0..self.amostras {
            let letra = (b'A' + (i % 26) as u8) as char;
            let sufixo: String = std::iter::repeat_n(letra, i % 10).collect();
            let texto = format!("TEXTO_VARIADO_{}{}", i, sufixo);
            let token = alvo
                .encrypt(texto.as_bytes())
                .map_err(|e| ErroAtaque::Falha(e.to_string()))?;
            let corpo = &token[alvo.prefixo().len()..];
            let decodificado = alvo
                .base64url_decode(corpo)
                .map_err(|e| ErroAtaque::Falha(e.to_string()))?;
            decodificados.push(decodificado);
        }

        let tamanho_minimo = decodificados.iter().map(|d| d.len()).min().unwrap_or(0);
        let mut posicoes_fixas: Vec<usize> = Vec::new();

        for i in 0..tamanho_minimo {
            let primeiro = decodificados[0][i];
            if decodificados.iter().all(|d| d[i] == primeiro) {
                posicoes_fixas.push(i);
            }
        }

        if !posicoes_fixas.is_empty() {
            let amostra: Vec<String> = posicoes_fixas
                .iter()
                .take(10)
                .map(|p| p.to_string())
                .collect();
            return Ok(ResultadoAtaque::new(
                &self.nome(),
                true,
                Severidade::Alta,
                &format!(
                    "{} posição(ões) de byte fixas em {} tokens: {}",
                    posicoes_fixas.len(),
                    self.amostras,
                    amostra.join(", ")
                ),
            ));
        }

        Ok(ResultadoAtaque::new(
            &self.nome(),
            false,
            Severidade::Info,
            &format!(
                "0 de {} posições fixas em {} tokens",
                tamanho_minimo, self.amostras
            ),
        ))
    }
}
